pennylane/transforms/qfunc_transforms.py

```diff
@@ -0,0 +1,306 @@
+
+    It can then be used to transform an existing qfunc:
+
+    >>> new_qfunc = my_qfunc_transform(0.6, 0.7)(old_qfunc)
+
+    It can also be used as a decorator:
+
+    .. code-block:: python
+
+        @qml.qfunc_transform
+        def my_transform(tape, x, y):
+            for op in tape.operations + tape.measurements:
+                if op.name == "CRX":
+                    wires = op.wires
+                    param = op.parameters[0]
+                    qml.RX(x * param, wires=wires[1])
+                    qml.RY(y * qml.math.sqrt(param), wires=wires[1])
+                    qml.CZ(wires=[wires[1], wires[0]])
+                else:
+                    op.queue()
+
+        @my_transform(0.6, 0.1)
+        def qfunc(x):
+            qml.Hadamard(wires=0)
+            qml.CRX(x, wires=[0, 1])
+
+    Let's use this qfunc to create a QNode, so that we can execute it on a quantum
+    device:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> qnode = qml.QNode(qfunc, dev)
+    >>> print(qml.draw(qnode)(2.5))
+     0: ──H───────────────────╭Z──┤
+     1: ──RX(1.5)──RY(0.158)──╰C──┤
+
+    Not only is the transformed qfunc fully differentiable, but the qfunc transform
+    parameters *themselves* are differentiable:
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit", wires=2)
+
+        def ansatz(x):
+            qml.Hadamard(wires=0)
+            qml.CRX(x, wires=[0, 1])
+
+        @qml.qnode(dev)
+        def circuit(param, transform_weights):
+            qml.RX(0.1, wires=0)
+
+            # apply the transform to the ansatz
+            my_transform(*transform_weights)(ansatz)(param)
+
+            return qml.expval(qml.PauliZ(1))
+
+    We can print this QNode to show that the qfunc transform is taking place:
+
+    >>> x = np.array(0.5, requires_grad=True)
+    >>> y = np.array([0.1, 0.2], requires_grad=True)
+    >>> print(qml.draw(circuit)(x, y))
+     0: ──RX(0.1)───H──────────╭Z──┤
+     1: ──RX(0.05)──RY(0.141)──╰C──┤ ⟨Z⟩
+
+    Evaluating the QNode, as well as the derivative, with respect to the gate
+    parameter *and* the transform weights:
+
+    >>> circuit(x, y)
+    0.9887793925354269
+    >>> qml.grad(circuit)(x, y)
+    (array(-0.02485651), array([-0.02474011, -0.09954244]))
+    """
+    if not callable(tape_transform):
+        raise ValueError(
+            "The qfunc_transform decorator can only be applied "
+            "to single tape transform functions."
+        )
+
+    if not isinstance(tape_transform, single_tape_transform):
+        tape_transform = single_tape_transform(tape_transform)
+
+    sig = inspect.signature(tape_transform)
+    params = sig.parameters
+
+    if len(params) > 1:
+
+        @functools.wraps(tape_transform)
+        def make_qfunc_transform(*targs, **tkwargs):
+            def wrapper(fn):
+                return _create_qfunc_internal_wrapper(fn, tape_transform, targs, tkwargs)
+
+            return wrapper
+
+    elif len(params) == 1:

```

Comment:
Nice :) This is the part that allows for the decorators to be specified without parens when there are no additional parameters?

Reply:
yep!